Using a SQL Magic Extension: Jupyter Notebook has a built-in extension called "SQL Magic" that allows you to write SQL queries directly in a notebook cell. To use this extension, you need to first install it by running `!pip install ipython-sql` in a notebook cell. Then, you can connect to your SQL database using the following code:

`!conda install -yc conda-forge ipython-sql`

`%load_ext sql`

`%sql postgresql://username:password@host:port/database_name`

In [1]:
%load_ext sql
%sql mysql://root:passwordhere@localhost/projects

'Connected: root@projects'

In [4]:
%sql SELECT * FROM hr LIMIT 5

 * mysql://root:***@localhost/projects
5 rows affected.


id,first_name,last_name,birthdate,gender,race,department,jobtitle,location,hire_date,termdate,location_city,location_state
00-0037846,Kimmy,Walczynski,06-04-91,Male,Hispanic or Latino,Engineering,Programmer Analyst I,Headquarters,1/20/2002,,Cleveland,Ohio
00-0041533,Ignatius,Springett,6/29/1984,Male,White,Business Development,Business Analyst,Headquarters,04-08-19,,Cleveland,Ohio
00-0045747,Corbie,Bittlestone,7/29/1989,Male,Black or African American,Sales,Solutions Engineer Manager,Headquarters,10-12-10,,Cleveland,Ohio
00-0055274,Baxy,Matton,9/14/1982,Female,White,Services,Service Tech,Headquarters,04-10-05,,Cleveland,Ohio
00-0076100,Terrell,Suff,04-11-94,Female,Two or More Races,Product Management,Business Analyst,Remote,9/29/2010,2029-10-29 06:09:38 UTC,Flint,Michigan


### Rename id column to emp_id

In [34]:
%%sql
ALTER TABLE hr
CHANGE COLUMN id emp_id VARCHAR(20) NULL

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Check data types of all columns

In [48]:
%%sql
DESCRIBE hr

 * mysql://root:***@localhost/projects
14 rows affected.


Field,Type,Null,Key,Default,Extra
emp_id,varchar(20),YES,,None,
first_name,text,YES,,None,
last_name,text,YES,,None,
birthdate,date,YES,,None,
gender,text,YES,,None,
race,text,YES,,None,
department,text,YES,,None,
jobtitle,text,YES,,None,
location,text,YES,,None,
hire_date,date,YES,,None,


### Change birthdate values to date

In this SQL code, we convert birthdate values from various formats to the standard `YYYY-MM-DD` format, handling both slash (`m/d/Y`) and dash (`m-d-y`) patterns and setting unmatched formats to NULL.

**In this code, we first check if the value contains a forward slash '/' using the LIKE operator. If it does, we assume the format is '%m/%d/%Y' and convert the value using the STR_TO_DATE() and DATE_FORMAT() functions to the '%Y-%m-%d' format. If it contains a dash '-', we assume the format is '%m-%d-%y' and convert the value to the '%Y-%m-%d' format. If the value does not match either format, we set the birthdate value to NULL.**

**Note that the DATE_FORMAT() function is used to convert the value to the '%Y-%m-%d' format, which is the standard MySQL date format. You can adjust the format string in the DATE_FORMAT() function to match your specific needs if you prefer a different date format.**

In [25]:
%%sql
UPDATE hr
SET birthdate = CASE
  WHEN birthdate LIKE '%/%' THEN DATE_FORMAT(STR_TO_DATE(birthdate, '%m/%d/%Y'), '%Y-%m-%d')
  WHEN birthdate LIKE '%-%' THEN DATE_FORMAT(STR_TO_DATE(birthdate, '%m-%d-%y'), '%Y-%m-%d')
  ELSE NULL
END;

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

In [47]:
%sql SELECT * FROM hr LIMIT 5

 * mysql://root:***@localhost/projects
5 rows affected.


emp_id,first_name,last_name,birthdate,gender,race,department,jobtitle,location,hire_date,termdate,location_city,location_state,age
00-0037846,Kimmy,Walczynski,1991-06-04,Male,Hispanic or Latino,Engineering,Programmer Analyst I,Headquarters,2002-01-20,None,Cleveland,Ohio,31
00-0041533,Ignatius,Springett,1984-06-29,Male,White,Business Development,Business Analyst,Headquarters,2019-04-08,None,Cleveland,Ohio,38
00-0045747,Corbie,Bittlestone,1989-07-29,Male,Black or African American,Sales,Solutions Engineer Manager,Headquarters,2010-10-12,None,Cleveland,Ohio,33
00-0055274,Baxy,Matton,1982-09-14,Female,White,Services,Service Tech,Headquarters,2005-04-10,None,Cleveland,Ohio,40
00-0076100,Terrell,Suff,1994-04-11,Female,Two or More Races,Product Management,Business Analyst,Remote,2010-09-29,2029-10-29,Flint,Michigan,28


### Change birthdate column datatype

In [28]:
# change data types
%sql ALTER TABLE hr MODIFY COLUMN birthdate DATE

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

In [60]:
%%sql
DESCRIBE hr

 * mysql://root:***@localhost/projects
13 rows affected.


Field,Type,Null,Key,Default,Extra
emp_id,varchar(20),YES,,None,
first_name,text,YES,,None,
last_name,text,YES,,None,
birthdate,date,YES,,None,
gender,text,YES,,None,
race,text,YES,,None,
department,text,YES,,None,
jobtitle,text,YES,,None,
location,text,YES,,None,
hire_date,date,YES,,None,


### Convert hire_date values to date

In [42]:
%%sql
UPDATE hr
SET hire_date = CASE
  WHEN hire_date LIKE '%/%' THEN DATE_FORMAT(STR_TO_DATE(hire_date, '%m/%d/%Y'), '%Y-%m-%d')
  WHEN hire_date LIKE '%-%' THEN DATE_FORMAT(STR_TO_DATE(hire_date, '%m-%d-%y'), '%Y-%m-%d')
  ELSE NULL
END

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Change hire_date column data type

In [44]:
%sql ALTER TABLE hr MODIFY COLUMN hire_date DATE

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Convert termdate values to date and remove time

In [54]:
%%sql
UPDATE hr
SET termdate = date(STR_TO_DATE(termdate, '%Y-%m-%d %H:%i:%s UTC'))
WHERE termdate IS NOT NULL AND termdate != ' '

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

In [165]:
%sql SELECT termdate FROM hr LIMIT 10

 * mysql://root:***@localhost/projects
10 rows affected.


termdate
None
None
None
None
2029-10-29
None
None
None
2008-12-05
None


### Convert termdate column to date

In [61]:
%sql ALTER TABLE hr MODIFY COLUMN termdate DATE

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Add age column

In [85]:
%%sql
ALTER TABLE hr ADD COLUMN age INT;
UPDATE hr SET age = TIMESTAMPDIFF(YEAR, birthdate, CURDATE());

 * mysql://root:***@localhost/projects
0 rows affected.
22214 rows affected.


[]

In [93]:
%%sql
SELECT 
  MIN(age) AS youngest,
  MAX(age) AS oldest
FROM hr

 * mysql://root:***@localhost/projects
1 rows affected.


youngest,oldest
-46,57


In [94]:
%%sql
SELECT COUNT(*) FROM hr WHERE age <18

 * mysql://root:***@localhost/projects
1 rows affected.


COUNT(*)
967


### Check Termdates in the future

In [125]:
%%sql
SELECT COUNT(*)
FROM hr
WHERE termdate > CURDATE()

 * mysql://root:***@localhost/projects
1 rows affected.


COUNT(*)
1599


In [149]:
%%sql
SELECT COUNT(*)
FROM hr
WHERE termdate = '0000-00-00'

 * mysql://root:***@localhost/projects
1 rows affected.


COUNT(*)
18285


## Questions

1. What is the gender breakdown of employees in the company?
2. What is the race/ethnicity breakdown of employees in the company?
3. What is the age distribution of employees in the company?
4. How many employees work at headquarters versus remote locations?
7. What is the distribution of job titles across the company?
8. Which department has the highest turnover rate?


In [96]:
%sql SHOW COLUMNS FROM hr

 * mysql://root:***@localhost/projects
14 rows affected.


Field,Type,Null,Key,Default,Extra
emp_id,varchar(20),YES,,None,
first_name,text,YES,,None,
last_name,text,YES,,None,
birthdate,date,YES,,None,
gender,text,YES,,None,
race,text,YES,,None,
department,text,YES,,None,
jobtitle,text,YES,,None,
location,text,YES,,None,
hire_date,date,YES,,None,


### 1. What is the gender breakdown of employees in the company?

In [30]:
%%sql
SELECT gender, COUNT(*) AS count
FROM hr
WHERE age >= 18
GROUP BY gender

 * mysql://root:***@localhost/projects
3 rows affected.


gender,count
Male,10794
Female,9876
Non-Conforming,577


### 2. What is the race/ethnicity breakdown of employees in the company?

In [98]:
%%sql
SELECT race, COUNT(*) AS count
FROM hr
WHERE age >= 18
GROUP BY race
ORDER BY count DESC

 * mysql://root:***@localhost/projects
7 rows affected.


race,count
White,6057
Two or More Races,3508
Black or African American,3443
Asian,3388
Hispanic or Latino,2402
American Indian or Alaska Native,1273
Native Hawaiian or Other Pacific Islander,1176


### 3. What is the age distribution of employees in the company?

In [46]:
%%sql
SELECT 
  MIN(age) AS youngest,
  MAX(age) AS oldest
FROM hr
WHERE age >= 18

 * mysql://root:***@localhost/projects
1 rows affected.


youngest,oldest
20,57


In [104]:
%%sql
SELECT FLOOR(age/10)*10 AS age_group, COUNT(*) AS count
FROM hr
WHERE age >= 18
GROUP BY FLOOR(age/10)*10;

 * mysql://root:***@localhost/projects
4 rows affected.


age_group,count
30,6170
40,5897
20,5680
50,3500


In [102]:
%%sql
SELECT 
  CASE 
    WHEN age >= 18 AND age <= 24 THEN '18-24'
    WHEN age >= 25 AND age <= 34 THEN '25-34'
    WHEN age >= 35 AND age <= 44 THEN '35-44'
    WHEN age >= 45 AND age <= 54 THEN '45-54'
    WHEN age >= 55 AND age <= 64 THEN '55-64'
    ELSE '65+' 
  END AS age_group, 
  COUNT(*) AS count
FROM 
  hr
WHERE 
  age >= 18
GROUP BY age_group
ORDER BY age_group;

 * mysql://root:***@localhost/projects
5 rows affected.


age_group,count
18-24,2709
25-34,6067
35-44,5978
45-54,5607
55-64,886


In [103]:
%%sql
SELECT 
  CASE 
    WHEN age >= 18 AND age <= 24 THEN '18-24'
    WHEN age >= 25 AND age <= 34 THEN '25-34'
    WHEN age >= 35 AND age <= 44 THEN '35-44'
    WHEN age >= 45 AND age <= 54 THEN '45-54'
    WHEN age >= 55 AND age <= 64 THEN '55-64'
    ELSE '65+' 
  END AS age_group, gender,
  COUNT(*) AS count
FROM 
  hr
WHERE 
  age >= 18
GROUP BY age_group, gender
ORDER BY age_group, gender;

 * mysql://root:***@localhost/projects
15 rows affected.


age_group,gender,count
18-24,Female,1235
18-24,Male,1408
18-24,Non-Conforming,66
25-34,Female,2854
25-34,Male,3051
25-34,Non-Conforming,162
35-44,Female,2686
35-44,Male,3134
35-44,Non-Conforming,158
45-54,Female,2675


### 4. How many employees work at headquarters versus remote locations?

In [106]:
%%sql
SELECT location, COUNT(*) as count
FROM hr
WHERE age >= 18
GROUP BY location;

 * mysql://root:***@localhost/projects
2 rows affected.


location,count
Headquarters,15992
Remote,5255


### 5. What is the distribution of job titles across the company?

In [110]:
%%sql
SELECT jobtitle, COUNT(*) AS count
FROM hr
WHERE age >= 18
GROUP BY jobtitle
ORDER BY count DESC;

### 6. Which department has the highest turnover rate?

In [43]:
%%sql
SELECT department, gender, COUNT(*) as count
FROM hr
WHERE age >= 18
GROUP BY department, gender
ORDER BY department, gender;

 * mysql://root:***@localhost/projects
38 rows affected.


department,gender,count
Accounting,Female,1462
Accounting,Male,1640
Accounting,Non-Conforming,90
Auditing,Female,23
Auditing,Male,27
Business Development,Female,722
Business Development,Male,798
Business Development,Non-Conforming,49
Engineering,Female,2983
Engineering,Male,3225


In [34]:
%%sql
SELECT department, ROUND(AVG(DATEDIFF(CURDATE(), termdate)/365),0) as avg_tenure
FROM hr
WHERE termdate <= CURDATE() AND termdate <> '0000-00-00' AND age >= 18
GROUP BY department

 * mysql://root:***@localhost/projects
13 rows affected.


department,avg_tenure
Engineering,8
Services,7
Human Resources,8
Business Development,8
Sales,7
Auditing,9
Training,8
Accounting,8
Research and Development,8
Product Management,8


## Summary of Findings

- There are more male employees
- White race is the most dominant while Native Hawaiian and American Indian are the least dominant.
- The youngest employee is 20 years old and the oldest is 57 years old
- 5 age groups were created (18-24, 25-34, 35-44, 45-54, 55-64). A large number of employees were between 25-34 followed by 35-44 while the smallest group was 55-64.
- A large number of employees work at the headquarters versus remotely.
- The average length of employment for terminated employees is around 7 years.
- The gender distribution across departments is fairly balanced but there are generally more male than female employees.
- The Marketing department has the highest turnover rate followed by Training. The least turn over rate are in the Research and development, Support and Legal departments.
- A large number of employees come from the state of Ohio.
- The net change in employees has increased over the years.
- The average tenure for each department is about 8 years with Legal and Auditing having the highest and Services, Sales and Marketing having the lowest.

## Limitations
- Some records had negative ages and these were excluded during querying(967 records). Ages used were 18 years and above.
- Some termdates were far into the future and were not included in the analysis(1599 records). The only term dates used were those less than or equal to the current date.